File: src/main/java/guru/drinkit/service/RecipeService.java
package guru.drinkit.service;

import java.io.IOException;
import java.util.List;

import guru.drinkit.common.Criteria;
import guru.drinkit.domain.Recipe;
import org.springframework.security.access.prepost.PreAuthorize;

public interface RecipeService {

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    Recipe save(Recipe recipe);

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    void delete(int id);

    List<Recipe> findAll();

    List<Recipe> findByCriteria(Criteria criteria);

    List<Recipe> findByRecipeNameContaining(String namePart);

    Recipe findById(int id);

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    void saveMedia(int recipeId, byte[] image, byte[] thumbnail) throws IOException;
}

File: src/main/java/guru/drinkit/service/RecipeService.kt
package guru.drinkit.service

import guru.drinkit.common.Criteria
import guru.drinkit.domain.Recipe
import org.springframework.security.access.prepost.PreAuthorize

interface RecipeService {

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    fun insert(recipe: Recipe): Recipe

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    fun update(recipe: Recipe): Recipe

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    fun delete(id: Int)

    fun findAll(): List<Recipe>

    fun findByCriteria(criteria: Criteria): List<Recipe>

    fun findByRecipeNameContaining(namePart: String): List<Recipe>

    fun findById(id: Int): Recipe?

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    fun saveMedia(recipeId: Int, image: ByteArray?, thumbnail: ByteArray?)
}
